keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,951 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The proposed enhancement adds a new `VolumeSource` to Kubernetes that supports OCI images and/or OCI artifacts.
+This allows users to package files and share them among containers in a pod without including them in the main image,
+thereby reducing vulnerabilities and simplifying image creation.
+
+While OCI images are well-supported by Kubernetes and CRI,
+extending support to OCI artifacts involves recognizing additional media types, implementing custom lifecycle management,
+and ensuring appropriate validation and security measures.
+
+## Motivation
+
+Supporting OCI images and artifacts directly as a `VolumeSource` allows
+Kubernetes to focus on OCI standards as well as allows to store and distribute
+any content using OCI registries. This allows the project to grow into use cases
+which go beyond running particular images.
+
+### Goals
+
+- Introduce a new `VolumeSource` type that allows mounting OCI images and/or artifacts.
+- Simplify the process of sharing files among containers in a pod.
+
+### Non-Goals
+
+- This proposal does not aim to replace existing `VolumeSource` types.
+- This proposal does not address other use cases for OCI objects beyond file sharing among containers in a pod.
+
+## Proposal
+
+We propose to add a new `VolumeSource` that supports OCI images and/or artifacts. This `VolumeSource` will allow users to mount an OCI object
+directly into a pod, making the files within the image accessible to the containers without the need to include them in the main image and to be able to host them in OCI compatible registries.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a Kubernetes user, I want to share a configuration file among multiple containers in a pod without including the file in my main image, so that I can
+minimize security risks and image size. I want to package this file in an OCI object to take advantage of OCI distribution.
+
+#### Story 2
+
+As a DevOps engineer, I want to package and distribute binary artifacts using OCI images and mount them directly into my Kubernetes pods,
```

Comment:
for the binary we need to make sure that the host architecture is matching. This is part of a plan, correct?

Reply:
:thought_balloon: I wonder if, _by default_, we should mount this kind of volume as `noexec`. I'd want that default.

Reply:
Yes, that should be done by the runtime on pull.

Reply:
Added a note about mounting with `noexec` within runtimes.

Reply:
We have a couple ways to make sure the artifact is "matching" wrt platform host arch/version etc. 

There is a new referrers api in the oci 1.1 spec that would allow us the opportunity to get a list of / filter artifacts that are referring to a platform image.. (if the registry and client supports the referrers api).  Rough steps .. pull platform index, match platform image, request artifacts referring to matched platform image, match requested artifact from this new k8s api.  We can discuss other patterns that do not rely on the referrers api on a per registry basis, and not all "1.0" oci registries support artifacts in the same way yet. 

We've had some discussions about a-priori pushing oci index content that includes artifacts ordered by platform. That would change up the ^ sequence to pulling the set of referred / tagged manifests and finding/matching the artifacts within that set. 

@sudo-bmitch